src/vm/threads.cpp

```diff
@@ -6551,6 +6559,106 @@ Thread *ThreadStore::GetThreadList(Thread *cursor)
     return GetAllThreadList(cursor, (Thread::TS_Unstarted | Thread::TS_Dead), 0);
 }
 
+const LONG ThreadStore::s_DeadThreadCountThresholdForGCTrigger = 25;
+const DWORD ThreadStore::s_DeadThreadGCTriggerPeriodMilliseconds = 1000 * 60 * 10;
+
+//
+// Dead thread count for GC trigger
+// - Interlocked operations are used, as a background GC thread may also be involved in reading/writing the value
+// - The thread store lock could theoreticaly be used, but it would add new GC suspension points and it is unclear that this is
+//   ok at these points
+//
+
+LONG ThreadStore::GetDeadThreadCountForGCTrigger()
+{
+    // Usage of this value should not assume a nonnegative value, see Increment and Decrement
+    return FastInterlockCompareExchange(&m_DeadThreadCountForGCTrigger, -1, -1);
```

Comment:
I just started with a safe change by default since the memory is accessed and modified from different threads. The purpose of a memory barrier here is to read a recent value so as not to miss a write of zero by a background GC thread some time back (probably don't need an interlocked read). I have now moved the determination of whether to do a GC and update of state into the lock (in TransferStartedThread), so that only one thread will trigger a GC. Also practically, a memory barrier would have been issued shortly before the read due to the lock, so I have removed the interlocked read / memory barrier and added explanation on why it is not necessary.